src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java

```diff
@@ -0,0 +1,121 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.plugins;
+
+import com.mysql.cj.Messages;
+import com.mysql.cj.conf.ConnectionUrl;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.log.Log;
+
+import com.amazonaws.services.secretsmanager.AWSSecretsManager;
+import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder;
+import com.amazonaws.services.secretsmanager.model.GetSecretValueRequest;
+import com.amazonaws.services.secretsmanager.model.GetSecretValueResult;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.sql.SQLException;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+public class AWSSecretsManagerPlugin implements IConnectionPlugin {
+  private final ICurrentConnectionProvider currentConnectionProvider;
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+  private final PropertySet propertySet;
+  private final ObjectMapper mapper = new ObjectMapper();
+  private final String region;
+  private final String secretId;
+
+  public AWSSecretsManagerPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.propertySet = propertySet;
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+    this.region = propertySet.getStringProperty(PropertyKey.secretsManagerRegion).getValue();
+    this.secretId = propertySet.getStringProperty(PropertyKey.secretsManagerSecretId).getValue();
+  }
+
+  @Override
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    AWSSecretsManager client  = AWSSecretsManagerClientBuilder.standard()
+        .withRegion(this.region)
+        .build();
+    GetSecretValueRequest getSecretValueRequest = new GetSecretValueRequest().withSecretId(this.secretId);
+    GetSecretValueResult getSecretValueResult;
+    String secret = "";
+    Properties updatedProperties = new Properties();
+    updatedProperties.putAll(connectionUrl.getOriginalProperties());
+
+    try {
+      getSecretValueResult = client.getSecretValue(getSecretValueRequest);
+      if (getSecretValueResult.getSecretString() != null) {
+        secret = getSecretValueResult.getSecretString();
+      }
+
+      JsonNode jsonObject = mapper.readTree(secret);
+      updatedProperties.put("username", jsonObject.get("username").asText());
+      updatedProperties.put("password", jsonObject.get("password").asText());
+    } catch (Exception e) {
+      throw new SQLException(Messages.getString("AWSSecretsManagerPlugin.1"));
+    }
+
+    ConnectionUrl newConnectionUrl = ConnectionUrl.getConnectionUrlInstance(connectionUrl.getDatabaseUrl(), updatedProperties);
```

Comment:
I believe we can use the same object instance

Reply:
I made a new ConnectionUrl instance because I wasn't able to change the connectionUrl properties directly.